### PR TITLE
SPOI-5542 #resolve #comment Prune the external dependencies from type…

### PR DESCRIPTION
**Pull request by siyuanh**
…graph and break the circle type

It is a blocker to schema, but it fix very rare cases as well and reduce the memory usage by typegraph.
So please cherrypick to 3.0 

Chandni, you can checkout my branch to unblock your work

### CODE REVIEW

File: engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java [hunk 1]
```diff
@@ -889,7 +895,8 @@ private void setTypes(JSONObject propJ, Type rawType, Map<Type, Type> typeReplac
           JSONArray jArray = new JSONArray();
           for (Type ttn : ((ParameterizedTypeNode) t).getActualTypeArguments()) {
             JSONObject objJ = new JSONObject();
-            setTypes(objJ, ttn, typeReplacement);
+            if(!stopRecursive)
+              setTypes(objJ, ttn, typeReplacement, visitedType);
```
Comment (chandnisingh) [inline engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java]: Missing braces

File: engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java [hunk 2]
```diff
@@ -926,20 +933,71 @@ private void setTypes(JSONObject propJ, Type rawType, Map<Type, Type> typeReplac
         propJ.put("uiType", UI_TYPE.LIST.getName());
         
         JSONObject jObj = new JSONObject();
-        setTypes(jObj, ((ArrayTypeNode)t).getActualArrayType(), typeReplacement);
+        if(!stopRecursive)
+          setTypes(jObj, ((ArrayTypeNode)t).getActualArrayType(), typeReplacement, visitedType);
```
Comment (chandnisingh) [inline engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java]: Missing braces

File: engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java [hunk 3]
```diff
@@ -926,20 +933,71 @@ private void setTypes(JSONObject propJ, Type rawType, Map<Type, Type> typeReplac
...
         propJ.put("itemType", jObj);
       }
       
       if(t instanceof TypeVariableNode){
         propJ.put("typeLiteral", ((TypeVariableNode)t).getTypeLiteral());
-        setTypes(propJ, ((TypeVariableNode)t).getRawTypeBound(), typeReplacement);
+        if(!stopRecursive)
+          setTypes(propJ, ((TypeVariableNode) t).getRawTypeBound(), typeReplacement, visitedType);
```
Comment (chandnisingh) [inline engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java]: Missing braces

File: engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java [hunk 4]
```diff
@@ -889,7 +895,8 @@ private void setTypes(JSONObject propJ, Type rawType, Map<Type, Type> typeReplac
           JSONArray jArray = new JSONArray();
           for (Type ttn : ((ParameterizedTypeNode) t).getActualTypeArguments()) {
             JSONObject objJ = new JSONObject();
-            setTypes(objJ, ttn, typeReplacement);
+            if(!stopRecursive)
```
Comment (davidyan74) [inline engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java]: Please always add space after if keyword.

File: engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java [hunk 5]
```diff
@@ -926,20 +933,71 @@ private void setTypes(JSONObject propJ, Type rawType, Map<Type, Type> typeReplac
         propJ.put("uiType", UI_TYPE.LIST.getName());
         
         JSONObject jObj = new JSONObject();
-        setTypes(jObj, ((ArrayTypeNode)t).getActualArrayType(), typeReplacement);
+        if(!stopRecursive)
```
Comment (davidyan74) [inline engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java]: Add space please

File: engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java [hunk 6]
```diff
@@ -926,20 +933,71 @@ private void setTypes(JSONObject propJ, Type rawType, Map<Type, Type> typeReplac
...
+          setTypes(jObj, ((ArrayTypeNode)t).getActualArrayType(), typeReplacement, visitedType);
         propJ.put("itemType", jObj);
       }
       
       if(t instanceof TypeVariableNode){
         propJ.put("typeLiteral", ((TypeVariableNode)t).getTypeLiteral());
-        setTypes(propJ, ((TypeVariableNode)t).getRawTypeBound(), typeReplacement);
+        if(!stopRecursive)
```
Comment (davidyan74) [inline engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java]: Add space please

**Comment by siyuanh** [comment]
Reformatted as requested

**Comment by davidyan74** [comment]
cherry picked to release-3.0

**Pull request merged**